f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py

```diff
@@ -134,186 +158,120 @@ def _assure_listeners_created(self, service):
 
         listeners = service["listeners"]
         loadbalancer = service["loadbalancer"]
-        networks = service["networks"]
+        networks = service.get("networks", list())
+        pools = service.get("pools", list())
+        l7policies = service.get("l7policies", list())
+        l7rules = service.get("l7policy_rules", list())
         bigips = self.driver.get_config_bigips()
 
         for listener in listeners:
-            pool = self.get_pool_by_id(
-                service, listener.get('default_pool_id', None))
-            svc = {"loadbalancer": loadbalancer,
-                   "listener": listener,
-                   "networks": networks}
+            error = False
+            if self._is_not_pending_delete(listener):
 
-            if pool:
-                svc['pool'] = pool
-            if listener['provisioning_status'] == \
-                    plugin_const.PENDING_UPDATE:
-                try:
-                    self.listener_builder.update_listener(svc, bigips)
-                except Exception as err:
-                    loadbalancer['provisioning_status'] = \
-                        plugin_const.ERROR
-                    listener['provisioning_status'] = plugin_const.ERROR
-                    raise f5_ex.VirtualServerUpdateException(err.message)
+                svc = {"loadbalancer": loadbalancer,
+                       "listener": listener,
+                       "pools": pools,
+                       "l7policies": l7policies,
+                       "l7policy_rules": l7rules,
+                       "networks": networks}
 
-            elif self._is_not_pending_delete(listener):
-                try:
-                    # create_listener() will do an update if VS exists
-                    self.listener_builder.create_listener(svc, bigips)
-                    listener['operating_status'] = \
-                        svc['listener']['operating_status']
-                except Exception as err:
+                # create_listener() will do an update if VS exists
+                error = self.listener_builder.create_listener(
+                    svc, bigips)
+
+                if error:
                     loadbalancer['provisioning_status'] = \
                         plugin_const.ERROR
                     listener['provisioning_status'] = plugin_const.ERROR
-                    raise f5_ex.VirtualServerCreationException(err.message)
-            self._set_status_as_active(listener)
+                else:
+                    listener['provisioning_status'] = plugin_const.ACTIVE
+                    listener['operating_status'] = lb_const.ONLINE
```

Comment:
Shouldn't he operating_status depends on the value of admin_state_up? The VS could have been created with admin_state_up = False.